### PR TITLE
Add `-Cbitcode-in-rlib`.

### PR DESCRIPTION
**Pull request by nnethercote**
This is a cut-down version of #70458 that gets the compile-time wins.

r? @alexcrichton

### CODE REVIEW

**Comment by nnethercote** [comment]
This will need https://github.com/rust-lang/cargo/pull/8134 to land first, and then I can add another commit that updates cargo. And #70729 also needs to land, and then I can remove the "XXX" commit.

**Comment by alexcrichton** [comment]
r=me once the relevant bits all land

**Comment by nnethercote** [comment]
@compiler-team: this is a cut-down alternative to #70458. That PR moved LLVM bitcode from a `.bc.z` file in the rlib into a bitcode section within the `.o` file in the rlib. Unfortunately, we discovered at the very end that [not all linkers can handle such a section](https://github.com/rust-lang/rust/pull/70458#issuecomment-615488333).

This PR is much simpler. It adds a new flag, `-Cbitcode-in-rlib`, which defaults to true for backwards compatibility. If you do specify it, the `.bc.z` is not put into the rlib. Cargo was taught to emit this flag when appropriate in https://github.com/rust-lang/cargo/pull/8134.

This goes back to the original idea from #66598, before it got pulled into a more expansive, but ultimately infeasible, direction. It attains all the performance benefits that #70458 did.

Normally a new `-C` flag would require compiler team approval and an FCP. However, because #70458 already went through that process and this PR is very much the same idea (just with a slightly different implementation), @alexcrichton, @Mark-Simulacrum, and I feel that a new FCP was not required.

**Comment by Mark-Simulacrum** [comment]
cc @rust-lang/compiler (the compiler-team org on github isn't related)

**Comment by nnethercote** [comment]
I should add some kind of test here, but I'm not sure what it would look like.

**Comment by nnethercote** [comment]
Uh, not sure why I'm getting this test failure:
```
error[E0053]: method `visit_union` has an incompatible type for trait
   --> src/tools/miri/src/helpers.rs:344:13
    |
344 |             fn visit_union(&mut self, v: MPlaceTy<'tcx, Tag>, fields: usize) -> InterpResult<'tcx> {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::num::NonZeroUsize`, found `usize`
    |
    = note: expected fn pointer `fn(&mut helpers::EvalContextExt::visit_freeze_sensitive::UnsafeCellVisitor<'ecx, 'mir, 'tcx, F>, rustc_mir::interpret::MPlaceTy<'_, _>, std::num::NonZeroUsize) -> std::result::Result<_, _>`
               found fn pointer `fn(&mut helpers::EvalContextExt::visit_freeze_sensitive::UnsafeCellVisitor<'ecx, 'mir, 'tcx, F>, rustc_mir::interpret::MPlaceTy<'tcx, _>, usize) -> std::result::Result<_, _>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0053`.
error: could not compile `miri`.
```
This PR doesn't touch anything that I would expect to have an effect like that.

**Comment by ehuss** [comment]
@nnethercote That error is from miri, which was broken by #70629.  When you update a submodule, it triggers the submodule builder for PRs.  But the miri error isn't the problem (miri failures are ignored at all times).  The PR breaks RLS:

```
error: Tool `rls` has regressed from test-pass to build-fail during beta week.
```

We'll need to update RLS.  I'll get started on a PR, though it will probably be better just to wait till after the beta branch.

**Comment by alexcrichton** [comment]
@nnethercote want to back out the Cargo update, land this, and we can get the Cargo update in later?

**Comment by nnethercote** [comment]
> @nnethercote want to back out the Cargo update, land this, and we can get the Cargo update in later?

I think it's easier to land it all together. Plus, having the Cargo update here means I can do a perf CI run to make sure everything is working:

@bors try @rust-timer queue

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -231,6 +231,16 @@ pub fn std_cargo(builder: &Builder<'_>, target: Interned<String>, cargo: &mut Ca
...
     }
+
+    // libstd must be built with bitcode so that the produced rlibs can be used
+    // for both LTO builds (which use bitcode) and non-LTO builds (which use
+    // object code).
+    //
+    // But we don't bother for the stage 0 compiler because it's never used
+    // with LTO.
+    if stage >= 1 {
+        cargo.rustflag("-Cbitcode-in-rlib=yes");
```
Comment (RalfJung) [inline src/bootstrap/compile.rs:242]: Somewhat confusing to see this explicitly set to the default value?

Comment (RalfJung) [reply src/bootstrap/compile.rs:242]: Oh this overwrite cargo which changes the default?

Comment (nnethercote) [reply src/bootstrap/compile.rs:242]: True, I will clarify the comment to explain that Cargo overrides the default, and we have to override Cargo's override.

**Comment by nnethercote** [comment]
Perf results are as expected.

@bors r=alexcrichton

File: src/bootstrap/compile.rs [hunk 2]
```diff
@@ -231,6 +231,18 @@ pub fn std_cargo(builder: &Builder<'_>, target: Interned<String>, cargo: &mut Ca
             }
         }
     }
+
+    // By default, rustc uses `-Cbitcode-in-rlib=yes`, and Cargo overrides that
+    // with `-Cbitcode-in-rlib=yes` for non-LTO builds. However, libstd must be
```
Comment (RalfJung) [inline src/bootstrap/compile.rs:236]: I suspect one of these "yes" should be "no".

Comment (nnethercote) [reply src/bootstrap/compile.rs]: True! Thanks, I have updated.

**Comment by nnethercote** [comment]
@bors p=1

Because of the performance win.

**Pull request merged**